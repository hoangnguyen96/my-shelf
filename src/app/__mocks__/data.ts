export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
