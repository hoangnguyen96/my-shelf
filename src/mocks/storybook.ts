export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};

export const mockSession = {
  user: {
    isAdmin: true,
    email: "admin@gmail.com",
    id: "3733403",
    name: "admin",
    image: "https://i.ibb.co/RHMqQGr/man-1.png",
  },
  expires: "2024-12-31T23:59:59.999Z",
};
