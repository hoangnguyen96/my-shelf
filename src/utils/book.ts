import { DEFAULT_LIMIT, TYPE_SEARCH } from "@app/constants";
import { BookType } from "@app/interface";

export const filterBooksOnShelf = (
  allBooks: BookType[],
  shelfBooks: string[]
) => allBooks.filter((item) => shelfBooks.includes(item.id));

export const filterBooksOnShelfByParams = (
  booksOnShelf: BookType[],
  type: string,
  value: string
) =>
  booksOnShelf.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const filterBooksFavorite = (
  allBooks: BookType[],
  favorites: string[]
) => allBooks.filter((item) => favorites.includes(item.id));

export const filterBooksFavoriteByParams = (
  booksFavorites: BookType[],
  type: string,
  value: string
) =>
  booksFavorites.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const dividePaginationBooks = (list: BookType[], limit?: number) => {
  return list.reduce((acc: BookType[][], _, i, self) => {
    if (!(i % (limit || DEFAULT_LIMIT))) {
      return [...acc, self.slice(i, i + (limit || DEFAULT_LIMIT))];
    }
    return acc;
  }, []);
};
