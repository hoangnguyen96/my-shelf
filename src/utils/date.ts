import { MONTHS, TIME_PERIODS } from "@app/constants";

export const formatDate = (date: Date) => {
  const hours = date.getHours();
  const minutes = date.getMinutes();
  const period = hours >= 12 ? TIME_PERIODS.PM : TIME_PERIODS.AM;
  const formattedHours = hours % 12 || 12;
  const formattedMinutes = minutes < 10 ? `0${minutes}` : minutes;

  return `${date.getDate()} ${MONTHS[date.getMonth()]} ${date.getFullYear()} ${formattedHours}:${formattedMinutes} ${period}`;
};
