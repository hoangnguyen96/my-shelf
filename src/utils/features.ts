import { DEFAULT_LIMIT, TYPE_SEARCH } from "@app/constants";
import { BookType } from "@app/models";
import {
  FieldErrors,
  FieldValues,
  Path,
  UseFormClearErrors,
} from "react-hook-form";
import { v4 as uuidv4 } from "uuid";

/**
 * Clear error message when the users typing
 */
export const clearErrorOnChange = <T extends FieldValues>(
  fieldName: Path<T>,
  errors: FieldErrors<T>,
  clearErrorFunc: UseFormClearErrors<T>
): void => {
  errors[fieldName]?.message && clearErrorFunc(fieldName);
};

export const generateSevenDigitUUID = () => {
  const uuid = uuidv4().replace(/\D/g, "");
  return uuid.slice(0, 7);
};

export const getTopBook = (data: BookType[], limit?: number) =>
  data
    .sort(
      (a, b) =>
        new Date(b.createdDate).getTime() - new Date(a.createdDate).getTime()
    )
    .slice(0, limit || DEFAULT_LIMIT);

export const formatDate = (date: Date) => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];

  const hours = date.getHours();
  const minutes = date.getMinutes();
  const period = hours >= 12 ? "PM" : "AM";
  const formattedHours = hours % 12 || 12;
  const formattedMinutes = minutes < 10 ? `0${minutes}` : minutes;

  return `${date.getDate()} ${months[date.getMonth()]} ${date.getFullYear()} ${formattedHours}:${formattedMinutes} ${period}`;
};

export const dividePaginationBooks = (list: BookType[]) => {
  return list.reduce((acc: BookType[][], _, i, self) => {
    if (!(i % 12)) {
      return [...acc, self.slice(i, i + 12)];
    }
    return acc;
  }, []);
};

export const getFirstPath = (value: string) => {
  if (!value) {
    return "";
  }

  const parts = value.split("/");
  const basePath = `/${parts[1]}`;

  return basePath;
};

export const getTwoPath = (value: string) => {
  if (!value) {
    return "";
  }

  const parts = value.split("/");
  const basePath = `/${parts[1]}/${parts[2]}`;

  return basePath;
};

export const filterBooksOnShelf = (
  allBooks: BookType[],
  shelfBooks: string[]
) => allBooks.filter((item) => shelfBooks.includes(item.id));

export const filterBooksOnShelfByParams = (
  booksOnShelf: BookType[],
  type: string,
  value: string
) =>
  booksOnShelf.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const filterBooksFavorite = (
  allBooks: BookType[],
  favorites: string[]
) => allBooks.filter((item) => favorites.includes(item.id));

export const filterBooksFavoriteByParams = (
  booksFavorites: BookType[],
  type: string,
  value: string
) =>
  booksFavorites.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const getListDataByTypeAndValue = (
  type: string,
  value: string,
  dataByParams: BookType[][],
  dataAll: BookType[][],
  pagination: number
) =>
  type && value ? dataByParams[pagination] || [] : dataAll[pagination] || [];

export const getDataByParams = (
  type: string,
  value: string,
  dataByParams: BookType[][],
  dataAll: BookType[][]
) => (type && value ? dataByParams : dataAll);
