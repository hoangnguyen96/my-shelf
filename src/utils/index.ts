export * from "./book";
export * from "./date";
export * from "./error";
export * from "./search";
export * from "./user";
export * from "./validators";
