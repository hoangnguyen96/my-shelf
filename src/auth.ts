import NextAuth from "next-auth";
import { authConfig } from "./auth.config";
import Credentials from "next-auth/providers/credentials";
import { z } from "zod";
import bcrypt from "bcryptjs";
import { MESSAGES } from "./constants";
import { getUserByEmail } from "@app/features/dashboard/actions";
import { User } from "./models";

export const { handlers, auth, signIn, signOut } = NextAuth({
  ...authConfig,
  secret: process.env.AUTH_SECRET,
  providers: [
    Credentials({
      name: "Credentials",
      credentials: {
        email: { label: "Email", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        if (!credentials) {
          return null;
        }

        const parsedCredentials = z
          .object({ email: z.string().email(), password: z.string().min(6) })
          .safeParse(credentials);

        if (parsedCredentials.success) {
          const { email, password } = parsedCredentials.data;
          const user = (await getUserByEmail(email)) as User;

          if (!user || !user.password) {
            throw new Error(MESSAGES.LOGIN_NOTFOUND);
          }

          const passwordsMatch = await bcrypt.compare(password, user.password);

          if (passwordsMatch)
            return {
              id: user.userId,
              name: user.username,
              isAdmin: user.isAdmin,
              image: user.avatar,
              email: user.email,
            };
        }

        throw new Error(MESSAGES.INVALID_CREDENTIALS);
      },
    }),
  ],
});
