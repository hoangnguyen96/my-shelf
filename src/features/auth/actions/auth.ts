"use server";

import { AuthError } from "next-auth";
import { MESSAGES } from "@app/constants";
import { User } from "@app/models";
import { signIn, signOut } from "@app/auth";

export const authenticate = async (formData: Partial<User>) => {
  try {
    await signIn("credentials", formData);
  } catch (error) {
    if (error instanceof AuthError) {
      switch (error.type) {
        case "CredentialsSignin":
          return MESSAGES.EMAIL_PASSWORD_INCORRECT;
        default:
          return MESSAGES.EMAIL_PASSWORD_INVALID;
      }
    }
    throw error;
  }
};

export const logout = async () => {
  await signOut();
};
