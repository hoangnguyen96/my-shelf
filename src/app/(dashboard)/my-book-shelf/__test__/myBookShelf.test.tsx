import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { useSession } from "next-auth/react";
import {
  getAllBook,
  getBookById,
  getUserById,
} from "@app/features/dashboard/actions";
import { DATA_BOOKS, DATA_USER } from "@app/mocks/data";
import * as utils from "@app/utils";
import MyBookShelfPage from "../page";
import { auth } from "@app/auth";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/auth", () => ({
  auth: jest.fn(),
}));

jest.mock("@app/features/dashboard/actions", () => ({
  getAllBook: jest.fn(),
  getUserById: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  filterBooksOnShelf: jest.fn(),
}));

describe("My Book Shelf", () => {
  const mockBooks = DATA_BOOKS.filter((item) =>
    DATA_USER[0].shelfBooks.includes(item.id)
  );
  const mockSession = {
    user: {
      isAdmin: true,
      email: "admin@gmail.com",
      id: "3733403",
      name: "admin",
      image: "https://i.ibb.co/RHMqQGr/man-1.png",
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (auth as jest.Mock).mockResolvedValue(mockSession);
    (getAllBook as jest.Mock).mockReturnValue({
      data: DATA_BOOKS,
    });
    (getUserById as jest.Mock).mockReturnValue({
      data: DATA_USER[0],
    });
    jest.spyOn(utils, "filterBooksOnShelf").mockReturnValue(mockBooks);
  });

  it("Should render correctly snapshot", async () => {
    const { container } = render(await MyBookShelfPage());

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it("Should render correctly snapshot when user has no shelfBooks", async () => {
    (getUserById as jest.Mock).mockResolvedValue({
      data: { ...DATA_USER[0], shelfBooks: null },
    });

    const { container } = render(await MyBookShelfPage());

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
