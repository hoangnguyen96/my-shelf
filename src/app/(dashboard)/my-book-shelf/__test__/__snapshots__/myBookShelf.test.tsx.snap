// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Book Shelf Should render correctly snapshot 1`] = `
<div>
  <div
    class="css-gh8r07"
  >
    <div
      class="css-bhtta9"
    >
      <div
        class="css-n736ek"
      >
        <img
          alt="Don't make me think"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="170"
          src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=256&q=75 2x"
          style="color: transparent;"
          width="123"
        />
        <p
          class="chakra-text css-0"
        >
          Don't Make Me Think
        </p>
        <div
          class="css-k008qs"
        >
          <p
            class="chakra-text css-0"
          >
            Steve Krug
            ,
          </p>
          <p
            class="chakra-text css-0"
          >
            2000
          </p>
        </div>
        <div
          class="css-k008qs"
        >
          <p
            class="chakra-text css-0"
          >
            4.5
            /
          </p>
          <p
            class="chakra-text css-c5z1pm"
          >
            5
          </p>
        </div>
      </div>
      <div
        class="css-1b6n4o1"
      >
        <div
          class="css-0"
        >
          <p
            class="chakra-text css-1r2f04i"
          >
            Borrowed on
          </p>
          <p
            class="chakra-text css-0"
          >
            3 Sep 2024 9:07 PM
          </p>
        </div>
        <button
          class="chakra-button css-16ltlly"
          data-testid="return-book"
          type="button"
        >
          Return
        </button>
      </div>
    </div>
  </div>
</div>
`;
